core/src/main/java/org/apache/calcite/sql/fun/SqlCastFunction.java

```diff
@@ -122,8 +128,50 @@ static SqlReturnTypeInference returnTypeInference(boolean safe) {
   /** Derives the type of "CAST(expression AS targetType)". */
   public static RelDataType deriveType(RelDataTypeFactory typeFactory,
       RelDataType expressionType, RelDataType targetType, boolean safe) {
-    return typeFactory.createTypeWithNullability(targetType,
-        expressionType.isNullable() || safe);
+    return createTypeWithNullabilityFromExpr(typeFactory, expressionType, targetType, safe);
+  }
+
+  public static RelDataType createTypeWithNullabilityFromExpr(RelDataTypeFactory typeFactory,
+      RelDataType expressionType, RelDataType targetType, boolean safe) {
+    boolean isn = expressionType.isNullable() || safe;
+
+    if (isCollection(expressionType)) {
+      RelDataType expressionElementType = expressionType.getComponentType();
+      RelDataType targetElementType = targetType.getComponentType();
+      RelDataType newElementType =
+          createTypeWithNullabilityFromExpr(typeFactory, expressionElementType, targetElementType, false);
```

Comment:
The third parameter should be `safe`?

Reply:
it should not. because the try_cast/safe_cast(different vendors, flink/spark former, mssql latter) will return null when it can not cast to targettype instead of throw exception like cast.

it only changes the outer nullable

Reply:
I'm not sure about this, if the inner type is `not null`, then safe cast should be `nullable` because the cast may produce `null`?

Reply:
yes @libenchao  
for exmpale:
a type is varchar not null
cast(a as timestamp)  throw exception
safe_cast/try_cast(a as timestamp) return null, the return type is varchar, which can be nullable.

Reply:
This is not what I meant to say. 

Take a example, `ARRAY<VARCHAR NOT NULL> NOT NULL`, safe_cast it to `ARRAY<INT NULLABLE> NULLABLE`, what should the result be? IMO, the inner type should be `NULLABLE` because it may produce `null`s.

Reply:
@libenchao why?
1. cast target type should not contains nullable like your example ARRAY<INT NULLABLE> NULLABLE
2. the array type and element type is different 
    array type nullable( outer nullable), the value is null
    element type nullable (inner nullable), the value is [null]

Reply:
because in safe mode, we'll produce `null`s, maybe a more concrete example explains this better: "['123', '456', 'abc']", if we cast it to `array<int>`, what should the result be? "[123, 456, null]", or "null" ? I don't know what other databases behaves in this case, but in my opinion, the former one is better. If this is also what other databases which provide safe cast be behaves, then the type for element type should also be nullable.

Reply:
fixed and add some test for try_cast and safe_cast